service/storage_service.cc

```diff
@@ -2032,8 +2032,40 @@ class topology_coordinator {
         slogger.info("raft topology: coordinator fiber found a node to work on id={} state={}", node.id, node.rs->state);
 
         switch (node.rs->state) {
-            case node_state::normal:
             case node_state::none: {
+                if (_topo_sm._topology.normal_nodes.empty()) {
+                    slogger.info("raft topology: skipping join node handshake for the first node in the cluster");
+                } else {
+                    auto validation_result = validate_joining_node(node);
+                    if (auto* reject = std::get_if<join_node_response_params::rejected>(&validation_result)) {
+                        // Find the IP first.
+                        auto ip = get_ip_for_joining_node_response(node.id);
+
+                        // Transition to left
+                        topology_mutation_builder builder(node.guard.write_timestamp());
+                        builder.with_node(node.id)
+                               .del("topology_request")
+                               .set("node_state", node_state::left);
+                        auto reason = ::format("bootstrap: node rejected");
+                        co_await update_topology_state(std::move(node.guard), {builder.build()}, reason);
+
+                        // Perform a single attempt to tell the node that it was removed.
+                        try {
+                            co_await respond_to_joining_node(node.id, ip, join_node_response_params{
+                                .response = std::move(validation_result),
+                            });
+                        } catch (...) {
+                            slogger.warn("raft topology: failed to send rejection response to {} ({}). Will not retry.", node.id, ip);
+                        }
+
+                        slogger.info("raft topology: rejected node moved to left state {} ({})", node.id, ip);
```

Comment:
I think it is fine (given that the joining node has a timeout and not just hangs). But how another coordinator can arrive to a different conclusion? It works on the same state.

Reply:
> But how another coordinator can arrive to a different conclusion? It works on the same state.

Right, if it works on the same state then it will arrive to the same conclusion.

A slightly different scenario then: topology coordinator sends reject response but fails to update `update_topology_state` because another node wrote a request to join. The topology coordinator retries and picks another node, bootstraps it, and only then looks at the node that was originally rejected, but the state is different. To be fair, I don't see how the state could change that the first node is rejected and then accepted, but validation might behave differently in the future.

Reply:
Yes, today features can only be enabled and thus if it was not OK to add a node it will stay this way.